### ISSUE TITLE
Create abstraction from spring-cloud-sleuth-correlation

### ISSUE THREAD

**Issue by spencergibb** [spring-cloud-incubator/spring-cloud-sleuth#2]
Allow implementations to plug into the log based solution (like s-c-s-zipkin) without having to recreate all of the touch points (instrumented classes, aop, etc...).

**Comment by spencergibb** [comment]
@marcingrzejszczak I'd like your feedback and questions on this (@mstine or @dsyer if you have the time as well).  I'll start to look at this tomorrow.

**Comment by mstine** [comment]
Pinging @b for feedback on this as well...

**Comment by marcingrzejszczak** [comment]
Can you elaborate more on what you would like to achieve cause from the description I don't get it too much? :P Isn't the autoconfiguration supposed to fix this?

**Comment by spencergibb** [comment]
@marcingrzejszczak how do you see the correlation id in the logs?  Do you need to change the log format?

**Comment by marcingrzejszczak** [comment]
Yes you have to. With MDC we set it and you have to customize the logging
message to include such a field (with Correlation-Id field name I guess)

**Comment by spencergibb** [comment]
![correlationid](https://cloud.githubusercontent.com/assets/594085/8070012/5ebf6bf0-0ecc-11e5-9059-72f053a32e71.png)

**Comment by spencergibb** [comment]
After perusing the 4finance correlation and the zipkin impl, I think I've started to form some thoughts about an abstraction.  The 4finance correlation id impl, it is fairly basic.  It put's a single UUID into the logs and passes it along through each request or thread.  I've started to think of their impl as the SLF4J impl.  The zipkin/brave impl is much more complex.

There are three parts to correlation:

1. starting/stopping a new trace
2. starting/stopping a new sub-trace (client tracing).
3. instrumenting packages to use the above.

__New Trace__
```
startTrace(C context);
stopTrace(C context);
```
The type of the context depends on how the request was started.  An http endpoint would look in an http request to see if it is part of an existing trace or not.  A new request starting from an `@Scheduled` method would have a different context.

__New Sub-trace__
```
startSubtrace(C context)
stopSubstrace(C context)
```
This may be better called client tracing.  This is where specific clients, such as `RestTemplate`, Spring MVC Controllers, Servlet Filters, etc... pass the trace data to request it constructs.  That may be via setting new http headers or passing values to a new thread.

__Instrumentation__
Instrumenting code can be accomplished in a couple of ways.  The 4finance code uses Spring AOP to intercept calls.  Other options include using existing extension points such as `RestTemplate` `RequestInterceptor`.  Another option is a delegate pattern.

@dsyer @mstine

**Comment by spencergibb** [comment]
@joshlong

**Comment by spencergibb** [comment]
PR #4 is the current work on this.

**Comment by spencergibb** [comment]
![sleuthtracing](https://cloud.githubusercontent.com/assets/594085/8387820/b1b55fa8-1c29-11e5-9a96-34cd692e494c.png)

Now with [traceId spanId] via #4

**Comment by spencergibb** [comment]
Closed via a51e92ad0bf72fb030b16fbb5b53ecadf59bb501
